lib/GenomeFileUtil/core/GenbankToGenome.py

```diff
@@ -282,97 +398,167 @@ def parse_genbank(self, file_path, params):
-                f"There were {self.defects['bad_parent_loc']} parent/child "
+        if genome_obj.defects['bad_parent_loc']:
+            genome_obj.genome_warnings.append(
+                f"There were {genome_obj.defects['bad_parent_loc']} parent/child "
                 "relationships that were not able to be determined. Some of "
                 "these may have splice variants that may be valid relationships."
             )
 
-        if self.defects['spoofed_genes']:
-            self.genome_warnings.append(warnings['spoofed_genome'].format(
-                self.defects['spoofed_genes']))
+        if genome_obj.defects['spoofed_genes']:
+            genome_obj.genome_warnings.append(
+                warnings['spoofed_genome'].format(genome_obj.defects['spoofed_genes'])
+            )
             genome['suspect'] = 1
 
-        if self.defects['not_trans_spliced']:
-            self.genome_warnings.append(warnings['genome_not_trans_spliced']
-                                        .format(self.defects['not_trans_spliced']))
+        if genome_obj.defects['not_trans_spliced']:
+            genome_obj.genome_warnings.append(
+                warnings['genome_not_trans_spliced'].format(
+                    genome_obj.defects['not_trans_spliced']
+                )
+            )
             genome['suspect'] = 1
 
-        if self.genome_warnings:
-            genome['warnings'] = self.genome_warnings
-        if self.genome_suspect:
+        if genome_obj.genome_warnings:
+            genome['warnings'] = genome_obj.genome_warnings
+        if genome_obj.genome_suspect:
             genome['suspect'] = 1
-        logging.info(f"Feature Counts: {genome['feature_counts']}")
-        return genome
 
-    def _save_assembly(self, genbank_file, params):
-        """Convert genbank file to fasta and sve as assembly"""
-        contigs = Bio.SeqIO.parse(genbank_file, "genbank")
-        assembly_id = f"{params['genome_name']}_assembly"
-        fasta_file = f"{self.cfg.sharedFolder}/{params['genome_name']}_assembly.fasta"
+        if params.get('genetic_code'):
+            genome['genetic_code'] = params['genetic_code']
+
+        logging.info(f"Feature Counts: {genome['feature_counts']}")
+        genome_obj.genome_data = genome
+
+    def _validate_existing_assembly(self, assembly_ref, genome_obj):
+        if not re.match("\d+\/\d+\/\d+", assembly_ref):
+            raise ValueError(f"Assembly ref: {assembly_ref} is not a valid format. Must"
+                                f" be in numerical <ws>/<object>/<version> format.")
+        ret = self.dfu.get_objects(
+            {'object_refs': [assembly_ref]}
+        )['data'][0]
+        if "KBaseGenomeAnnotations.Assembly" not in ret['info'][2]:
+            raise ValueError(f"{assembly_ref} is not a reference to an assembly")
+        unmatched_ids = list()
+        unmatched_ids_md5s = list()
+        for current_contig in genome_obj.contig_seq.keys():
```

Comment:
This seems to imply that the contigs are in memory which will almost certainly be a problem with thousands of genomes

Reply:
ah, this function is only for those genomes that `use_existing_assembly`, which are barely used in the codebase.
And I assume in collections we are only considering the following 3 cases for now:
1. assembly only
2. genome only
3. genome and then assembly

Reply:
> Tian's current code can support case 1 even we don't think we'll need it.

https://github.com/kbase/collections/pull/656#discussion_r1500185583

> case 2 I mean upload genome, and yes, internally it will either point to an existing assembly or upload assembly.

AFAIK we can only support uploading an assembly and genome at the same time, so case 2a) above and case 3 aren't going to be supported


> Under normal circumstances, _validate_existing_assembly will not be called, unless use_existing_assembly is provided.

But is that the only place contig_seq is populated?

Reply:
contig_seq is populated in `_get_contigs_and_extra_info`, which are used in both `_load_contigs_for_and_validate_existing_assemblies` and `_save_assemblies` functions.

Reply:
Ok, so is it populated in the code that's executed when you upload genbank files without providing an existing assembly?

Reply:
The answer is that all the contigs from all genomes are in memory at once for a regular upload, which is probably not great. I think a couple things need to happen (probably not in this PR, this review has gone on long enough):

* A general read through of the call tree looking at whats in memory at any given time and whether that could cause problems
* See if there's a way to load the contigs, do any necessary processing, and then drop them. You don't need them for saving the assembly or genome objects so it seems like that should be possible

Reply:
Hi, my understanding is contig_seq will be used in the _parse_genbank function to produce dicts that will update genome.data(required for upload). So I think we can drop them once the update is done. 

Would like to delay this to a future PR as well.

Reply:
Does that mean that per thread only 1 config_seq will be in memory at a time, once things are parallelized?

Reply:
IIUC, that depends on number of worker processes we decide to parallelize. I am not sure about thread as a process can have anywhere from one thread to many. Ideally, if we specify the number of worker processes same as CPU counts, then each work process should have 1 config_seq.